src/main/java/org/javarosa/core/model/TriggerableDag.java

```diff
@@ -270,144 +291,148 @@ public void copyItemsetAnswer(FormInstance mainInstance, EvaluationContext evalC
-    public void finalizeTriggerables(FormInstance mainInstance, EvaluationContext evalContext) throws IllegalStateException {
-        //
-        // DAGify the triggerables based on dependencies and sort them so that
-        // triggerables come only after the triggerables they depend on
-        //
-        List<QuickTriggerable> vertices = new ArrayList<>(unorderedTriggerables);
-        triggerablesDAG.clear();
-        List<QuickTriggerable[]> partialOrdering = new ArrayList<>();
-        Set<QuickTriggerable> newDestinationSet = new HashSet<>();
-        for (QuickTriggerable qt : vertices) {
-            Set<QuickTriggerable> deps = new HashSet<>();
-            newDestinationSet.clear();
-            fillTriggeredElements(mainInstance, evalContext, qt, deps, newDestinationSet);
-
-            if (deps.contains(qt))
-                throwCyclesInDagException(deps);
-
-            if (qt.canCascade())
-                for (QuickTriggerable qu : deps) {
-                    QuickTriggerable[] edge = {qt, qu};
-                    partialOrdering.add(edge);
-                }
+    public void finalizeTriggerables(FormInstance mainInstance) throws IllegalStateException {
+        List<QuickTriggerable[]> edges = getDagEdges(allTriggerables, triggerablesPerTrigger);
+        triggerablesDAG = buildDag(allTriggerables, edges);
+        repeatConditionsPerTargets = getRepeatConditionsPerTargets(mainInstance, triggerablesDAG);
+    }
 
-            qt.setImmediateCascades(deps);
+    private static Map<TreeReference, QuickTriggerable> getRepeatConditionsPerTargets(FormInstance mainInstance, Set<QuickTriggerable> triggerables) {
+        Map<TreeReference, QuickTriggerable> repeatConditionsPerTargets = new HashMap<>();
+        for (QuickTriggerable triggerable : triggerables)
+            if (triggerable.isCondition())
+                for (TreeReference target : triggerable.getTargets())
+                    if (mainInstance.getTemplate(target) != null)
+                        repeatConditionsPerTargets.put(target, triggerable);
+        return repeatConditionsPerTargets;
+    }
+
+    /**
+     * Returns the list of edges in the DAG that can be built from the provided
+     * vertices.
+     * <p>
+     * This method has side-effects:
+     * <ul>
+     *     <li>Throws IllegalStateException when cycles are detected, either due
+     *     to a self-reference or more complex cycle chains</li>
+     *     <li>Builds a cache of immediate cascades of each vertex, meaning that
+     *     we will remember the dependant vertices without having to traverse
+     *     the DAG again</li>
+     * </ul>
+     */
+    private static List<QuickTriggerable[]> getDagEdges(Set<QuickTriggerable> vertices, Map<TreeReference, Set<QuickTriggerable>> triggerIndex) {
+        List<QuickTriggerable[]> edges = new ArrayList<>();
+        for (QuickTriggerable source : vertices) {
+            // Compute the set of direct children triggerables of this source vertex
```

Comment:
Should this say "direct dependent" instead of "direct children"? In the case of relevance on a group, the references of all the nodes inside the group (and the groups inside the group, etc) will be included, right? All of these descendants are direct dependents of the containing group because their relevance is directly affected by the parent group's relevance.

Reply:
Ugh, language is hard. I think the comment is misleading by using the word "children". We should  stick with DAG terminology. I'll review this and try to improve it with your suggestions.